File: frontend/display_app/lib/data/catalog_repository.dart
import '../models/category.dart';
import '../models/product.dart';
import 'mock_catalog.dart';

/// Contract for catalog access. The UI depends only on this interface; the
/// mock implementation is swapped for an HTTP-backed one when the backend
/// clothes API is ready — no UI refactoring required.
abstract interface class CatalogRepository {
  Future<List<Category>> categories();
  Future<List<Product>> products({String? categoryId, String? query});
  Future<Product?> productById(String id);
}

/// In-memory implementation backed by [MockCatalog]. Simulates async latency
/// so loading states are exercised exactly as they will be against the network.
class MockCatalogRepository implements CatalogRepository {
  const MockCatalogRepository();

  static const Duration _latency = Duration(milliseconds: 350);

  @override
  Future<List<Category>> categories() async {
    await Future<void>.delayed(_latency);
    return MockCatalog.categories;
  }

  @override
  Future<List<Product>> products({String? categoryId, String? query}) async {
    await Future<void>.delayed(_latency);
    final String? q = query?.trim().toLowerCase();
    return MockCatalog.products.where((Product p) {
      final bool matchesCategory = categoryId == null || p.categoryId == categoryId;
      final bool matchesQuery = q == null ||
          q.isEmpty ||
          p.name.toLowerCase().contains(q) ||
          p.brand.toLowerCase().contains(q) ||
          p.aiHighlights.any((String h) => h.toLowerCase().contains(q));
      return matchesCategory && matchesQuery;
    }).toList();
  }

  @override
  Future<Product?> productById(String id) async {
    await Future<void>.delayed(_latency);
    return MockCatalog.products.where((Product p) => p.id == id).firstOrNull;
  }
}

File: frontend-rn/display_app/src/data/catalogRepository.ts
import { Category } from '../models/category';
import { Product } from '../models/product';
import { MockCatalog } from './mockCatalog';

/// Contract for catalog access. The UI depends only on this interface. Ported
/// from the Flutter `CatalogRepository`.
export interface CatalogRepository {
  categories(): Promise<Category[]>;
  products(opts?: { categoryId?: string; query?: string }): Promise<Product[]>;
  productById(id: string): Promise<Product | undefined>;
}

const delay = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/// In-memory implementation backed by MockCatalog with simulated latency.
export class MockCatalogRepository implements CatalogRepository {
  private static readonly latency = 350;

  async categories(): Promise<Category[]> {
    await delay(MockCatalogRepository.latency);
    return MockCatalog.categories;
  }

  async products(opts?: { categoryId?: string; query?: string }): Promise<Product[]> {
    await delay(MockCatalogRepository.latency);
    const q = opts?.query?.trim().toLowerCase();
    return MockCatalog.products.filter((p) => {
      const matchesCategory = opts?.categoryId == null || p.categoryId === opts.categoryId;
      const matchesQuery =
        q == null ||
        q.length === 0 ||
        p.name.toLowerCase().includes(q) ||
        p.brand.toLowerCase().includes(q) ||
        p.aiHighlights.some((h) => h.toLowerCase().includes(q));
      return matchesCategory && matchesQuery;
    });
  }

  async productById(id: string): Promise<Product | undefined> {
    await delay(MockCatalogRepository.latency);
    return MockCatalog.products.find((p) => p.id === id);
  }
}
